# `analyses` Subdirectory Structure

This subdirectory hosts the output of all analyses regarding *Z. marina* response to *L. zosterae* infection. Every folder houses all necessary files for a discrete analysis step in a larger pipeline. Folders are named after the program used to analyze data.

The following pipelines are represented in this subdirectory:

[EdgeR](https://github.com/eimd-2019/project-EWD-transcriptomics/tree/master/analyses/EdgeR): Differential expression analysis of transcriptomic data
- [GRASS_START_GENE_MEE_Contrasts_CAB_7_10_Zosteraonly_at.R](https://github.com/eimd-2019/project-EWD-transcriptomics/tree/master/analyses/EdgeR/GRASS_START_GENE_MEE_Contrasts_CAB_7_10_Zosteraonly_at.R): R script to identify differentially expressed genes in *Z. marina*
- [GRASS_START_GENE_MEE_Contrasts_CAB_7_10_nonZostera_at.R](https://github.com/eimd-2019/project-EWD-transcriptomics/blob/master/analyses/EdgeR/GRASS_START_GENE_MEE_Contrasts_CAB_7_10_nonZostera_at.R): R script to identify differentially expressed genes in non-*Zostera spp.*.
- [EXP.CON.Zostera.ALL.txt](https://github.com/eimd-2019/project-EWD-transcriptomics/blob/master/analyses/EdgeR/EXP.CON.Zostera.ALL.txt): Output with all *Z. marina* genes used in `edgeR` analysis and associated significance metrics
- [EXP.CON.NonZostera.ALL.txt](https://github.com/eimd-2019/project-EWD-transcriptomics/blob/master/analyses/EdgeR/EXP.CON.NonZostera.ALL.txt): Output with all non-*Zostera spp.* genes used in `edgeR` analysis and associated significance metrics
- [DE.EXP.CON.FDR.Z.Annot.txt](https://github.com/eimd-2019/project-EWD-transcriptomics/blob/master/analyses/EdgeR/DE.EXP.CON.FDR.Z.Annot.txt): Differentially expressed *Z. marina* genes and associated significance metrics
- [DE.EXP.CON.FDR.nZ.Annot.txt](https://github.com/eimd-2019/project-EWD-transcriptomics/blob/master/analyses/EdgeR/DE.EXP.CON.FDR.nZ.Annot.txt): Differentially expressed non-*Zostera spp.* genes and associated significance metrics

[GO-MWU](https://github.com/eimd-2019/project-EWD-transcriptomics/tree/master/analyses/GO-MWU): Gene enrichment analysis derived from [this pipeline](https://github.com/z0on/GO_MWU).
- [2019-07-11-Gene-Enrichment-with-GO-MWU.Rmd](https://github.com/eimd-2019/project-EWD-transcriptomics/blob/master/analyses/GO-MWU/2019-07-11-Gene-Enrichment-with-GO-MWU.Rmd): R Markdown file used to format GO-MWU inputs. Associated output found below
	- [2019-07-11-Zostera-GO-Annotations-Table.tab](https://github.com/eimd-2019/project-EWD-transcriptomics/blob/master/analyses/GO-MWU/2019-07-11-Zostera-GO-Annotations-Table.tab): *Z. marina* GO annotations. This tab-delimited has gene IDs and all associated GOterms separated by semicolons.
	- [2019-07-11-Zostera-Table-of-Significance-Measures.csv](https://github.com/eimd-2019/project-EWD-transcriptomics/blob/master/analyses/GO-MWU/2019-07-11-Zostera-Table-of-Significance-Measures.csv):*Z. marina* table of continuous significance measures. This .csv file has all gene IDs tested in `edgeR`  and signed log p-values.
	- [2019-07-11-nonZostera-GO-Annotations-Table.tab](https://github.com/eimd-2019/project-EWD-transcriptomics/blob/master/analyses/GO-MWU/2019-07-11-nonZostera-GO-Annotations-Table.tab): non-*Zostera spp.* GO annotations.
	- [2019-07-11-nonZostera-Table-of-Significance-Measures.csv](https://github.com/eimd-2019/project-EWD-transcriptomics/blob/master/analyses/GO-MWU/2019-07-11-nonZostera-Table-of-Significance-Measures.csv): non-*Zostera spp.* table of continuous significance measures.
- Files necessary to run GO-MWU downloaded from GO-MWU repository
	- [go.obo](https://github.com/eimd-2019/project-EWD-transcriptomics/blob/master/analyses/GO-MWU/go.obo): Gene ontology database
	- [gomwu_a.pl](https://github.com/eimd-2019/project-EWD-transcriptomics/blob/master/analyses/GO-MWU/gomwu_a.pl) and [gomwu_b.pl](https://github.com/eimd-2019/project-EWD-transcriptomics/blob/master/analyses/GO-MWU/gomwu_b.pl): Perl scripts
	- [gomwu.functions.R](https://github.com/eimd-2019/project-EWD-transcriptomics/blob/master/analyses/GO-MWU/gomwu.functions.R): Source file for R code
- [Zostera-GO_MWU.R](https://github.com/eimd-2019/project-EWD-transcriptomics/blob/master/analyses/GO-MWU/Zostera-GO_MWU.R): R Script used to run GO-MWU analysis on *Z. marina* genes. Associated output found below
	- [dissim_BP_2019-07-11-Zostera-GO-Annotations-Table.tab](https://github.com/eimd-2019/project-EWD-transcriptomics/blob/master/analyses/GO-MWU/dissim_BP_2019-07-11-Zostera-GO-Annotations-Table.tab): Dissimilarity matrix used for GO-MWU BP test
	- [dissim_CC_2019-07-11-Zostera-GO-Annotations-Table.tab](https://github.com/eimd-2019/project-EWD-transcriptomics/blob/master/analyses/GO-MWU/dissim_CC_2019-07-11-Zostera-GO-Annotations-Table.tab): Dissimilarity matrix used for GO-MWU CC test
	- [dissim_MF_2019-07-11-Zostera-GO-Annotations-Table.tab](https://github.com/eimd-2019/project-EWD-transcriptomics/blob/master/analyses/GO-MWU/dissim_MF_2019-07-11-Zostera-GO-Annotations-Table.tab): Dissimilarity matrix used for GO-MWU MF test
	- [BP_2019-07-11-Zostera-Table-of-Significance-Measures.csv](https://github.com/eimd-2019/project-EWD-transcriptomics/blob/master/analyses/GO-MWU/BP_2019-07-11-Zostera-Table-of-Significance-Measures.csv): GO-MWU p-values for BP gene ontology terms associated with individual genes
    - [CC_2019-07-11-Zostera-Table-of-Significance-Measures.csv](https://github.com/eimd-2019/project-EWD-transcriptomics/blob/master/analyses/GO-MWU/CC_2019-07-11-Zostera-Table-of-Significance-Measures.csv): GO-MWU p-values for CC gene ontology terms associated with individual genes
    - [MF_2019-07-11-Zostera-Table-of-Significance-Measures.csv](https://github.com/eimd-2019/project-EWD-transcriptomics/blob/master/analyses/GO-MWU/MF_2019-07-11-Zostera-Table-of-Significance-Measures.csv): GO-MWU p-values for MF gene ontology terms associated with individual genes
    - [MWU_BP_2019-07-11-Zostera-Table-of-Significance-Measures.csv](https://github.com/eimd-2019/project-EWD-transcriptomics/blob/master/analyses/GO-MWU/MWU_BP_2019-07-11-Zostera-Table-of-Significance-Measures.csv): GO-MWU p-values for BP gene ontology terms associated with parent terms. This is the final GO-MWU output
    - [MWU_CC_2019-07-11-Zostera-Table-of-Significance-Measures.csv](https://github.com/eimd-2019/project-EWD-transcriptomics/blob/master/analyses/GO-MWU/MWU_CC_2019-07-11-Zostera-Table-of-Significance-Measures.csv): GO-MWU p-values for CC gene ontology terms associated with parent terms. This is the final GO-MWU output
    - [MWU_MF_2019-07-11-Zostera-Table-of-Significance-Measures.csv](https://github.com/eimd-2019/project-EWD-transcriptomics/blob/master/analyses/GO-MWU/MWU_MF_2019-07-11-Zostera-Table-of-Significance-Measures.csv): GO-MWU p-values for MF gene ontology terms associated with parent terms. This is the final GO-MWU output
    - [MWU_MF_SIGONLY_2019-07-11-Zostera-Table-of-Significance-Measures.csv](https://github.com/eimd-2019/project-EWD-transcriptomics/blob/master/analyses/GO-MWU/MWU_MF_SIGONLY_2019-07-11-Zostera-Table-of-Significance-Measures.csv): Significantly enriched MF terms from GO-MWU
- [nonZostera-GO_MWU.R](https://github.com/eimd-2019/project-EWD-transcriptomics/blob/master/analyses/GO-MWU/nonZostera-GO_MWU.R):. R Script used to run GO-MWU analysis on non-*Zostera spp* genes. Associated output found below
	- [dissim_BP_2019-07-11-nonZostera-GO-Annotations-Table.tab](https://github.com/eimd-2019/project-EWD-transcriptomics/blob/master/analyses/GO-MWU/dissim_BP_2019-07-11-nonZostera-GO-Annotations-Table.tab): Dissimilarity matrix used for GO-MWU BP test
	- [dissim_CC_2019-07-11-nonZostera-GO-Annotations-Table.tab](https://github.com/eimd-2019/project-EWD-transcriptomics/blob/master/analyses/GO-MWU/dissim_CC_2019-07-11-nonZostera-GO-Annotations-Table.tab): Dissimilarity matrix used for GO-MWU CC test
	- [dissim_MF_2019-07-11-nonZostera-GO-Annotations-Table.tab](https://github.com/eimd-2019/project-EWD-transcriptomics/blob/master/analyses/GO-MWU/dissim_MF_2019-07-11-nonZostera-GO-Annotations-Table.tab): Dissimilarity matrix used for GO-MWU MF test
	- [BP_2019-07-11-nonZostera-Table-of-Significance-Measures.csv](https://github.com/eimd-2019/project-EWD-transcriptomics/blob/master/analyses/GO-MWU/BP_2019-07-11-nonZostera-Table-of-Significance-Measures.csv): GO-MWU p-values for BP gene ontology terms associated with individual genes
    - [CC_2019-07-11-nonZostera-Table-of-Significance-Measures.csv](https://github.com/eimd-2019/project-EWD-transcriptomics/blob/master/analyses/GO-MWU/CC_2019-07-11-nonZostera-Table-of-Significance-Measures.csv): GO-MWU p-values for CC gene ontology terms associated with individual genes
    - [MF_2019-07-11-nonZostera-Table-of-Significance-Measures.csv](https://github.com/eimd-2019/project-EWD-transcriptomics/blob/master/analyses/GO-MWU/MF_2019-07-11-nonZostera-Table-of-Significance-Measures.csv): GO-MWU p-values for MF gene ontology terms associated with individual genes
    - [MWU_BP_2019-07-11-nonZostera-Table-of-Significance-Measures.csv](https://github.com/eimd-2019/project-EWD-transcriptomics/blob/master/analyses/GO-MWU/MWU_BP_2019-07-11-nonZostera-Table-of-Significance-Measures.csv): GO-MWU p-values for BP gene ontology terms associated with parent terms. This is the final GO-MWU output
    - [MWU_CC_2019-07-11-nonZostera-Table-of-Significance-Measures.csv](https://github.com/eimd-2019/project-EWD-transcriptomics/blob/master/analyses/GO-MWU/MWU_CC_2019-07-11-nonZostera-Table-of-Significance-Measures.csv): GO-MWU p-values for CC gene ontology terms associated with parent terms. This is the final GO-MWU output
    - [MWU_MF_2019-07-11-nonZostera-Table-of-Significance-Measures.csv](https://github.com/eimd-2019/project-EWD-transcriptomics/blob/master/analyses/GO-MWU/MWU_MF_2019-07-11-nonZostera-Table-of-Significance-Measures.csv): GO-MWU p-values for MF gene ontology terms associated with parent terms. This is the final GO-MWU output