---
title: "Gene Expression Heatmaps"
author: "Yaamini Venkataraman"
date: "7/15/2019"
output: github_document
---

In this R Markdown document, I will create heatmaps for *Z. marina* and *L. zosterae* gene expression.

# Set up R Markdown document

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Install packages

```{r}
#install.packages("pheatmap") #Uncomment if you need to install this package
require(pheatmap)
```

```{r}
#install.packages("dplyr") #Uncomment if you need to install this package
require(dplyr)
```

# Obtain session information

```{r}
sessionInfo()
```

# *Z. marina* heatmaps

## Create color scheme

```{r}
RColorBrewer::display.brewer.all() #Look at RColorBrewer palettes. I'm looking for something that will still make sense for someone with red-green confusion.
```

```{r}
heatmapGreen <- RColorBrewer::brewer.pal(9, "Greens") #Save green color palette from RColorBrewer
```

## All genes

Using all genes in a heatmap will be noisy, and probably not that interesting since the changes will be hard to see.

*Verdict: I was right.*

```{r}
Zostera.GE<-read.delim("EXP.CON.Zostera.logCPM.txt", header=TRUE) #Import logCPM file
head(Zostera.GE) #Confirm import
```

```{r}
rownames(Zostera.GE) <- Zostera.GE$GeneID #Set gene ID as row names
Zostera.GE.heatmap <- Zostera.GE[,-1] #Remove redundant column
head(Zostera.GE.heatmap) #Confirm changes
```

```{r}
pheatmap(Zostera.GE.heatmap, cluster_row = TRUE, cluster_cols = FALSE, clustering_distance_rows = "euclidean", clustering_method = "average", show_rownames = FALSE, show_colnames = TRUE, treeheight_col = 80, legend = TRUE, color = heatmapGreen, fontsize_col = 20, fontsize_row = 12) #Create heatmap. Cluster rows, but not columns, Use euclidean distances and average method for clustering. Do not show rownames but show column names. Adjust tree height, add a legend, and specify colors and font sizes.
```

## Differentially expressed genes

*Verdict: This is definitely the best looking one.*

```{r}
zosteraDEG <- read.delim("DE.EXP.CON.FDR.Z.Annot.M.txt") #Import modified file for DEG
head(zosteraDEG) #Confirm import
```

```{r}
zosteraDEG.heatmap <- merge(x = Zostera.GE, y = zosteraDEG, by = "GeneID") #Merge files by GeneID
rownames(zosteraDEG.heatmap) <- zosteraDEG.heatmap$GeneID #Set gene IDs as rownames
zosteraDEG.heatmap <- zosteraDEG.heatmap[,c(2:16)] #Only keep columns with logCPM for each sample
head(zosteraDEG.heatmap) #Confirm changes
```

```{r}
#pdf("2019-07-15-Zostera-Differentially-Expressed-Genes-Heatmap.pdf", width = 11, height = 8.5) #Uncomment and run if you want to save the plot
pheatmap(zosteraDEG.heatmap, cluster_row = TRUE, cluster_cols = FALSE, clustering_distance_rows = "euclidean", clustering_method = "average", show_rownames = FALSE, show_colnames = TRUE, treeheight_col = 80, legend = TRUE, color = heatmapGreen, fontsize_col = 20, fontsize_row = 12) #Create heatmap. Cluster rows, but not columns, Use euclidean distances and average method for clustering. Do not show rownames but show column names. Adjust tree height, add a legend, and specify colors and font sizes.
#dev.off() #Uncomment and run after running plotting code above
```

## Differentially expressed genes with GO groupings

I've had `pheatmap` cluster rows for me, but I can also organize rows by gene ontology grouping before creating a heatmap. I will try this out.

*Verdict: Since the heatmap looks better if you cluster rows, there is no point in organzing by groupings the. DO NOT RUN THE CODE IN THIS SECTION YOU WILL GET GARBAGE.*

```{r}
zosteraDEGroupings <- read.csv("../GO-MWU/DE-GO-MWU/2019-07-15-Zostera-DEG-GoGroupings.csv", header = TRUE) #Import file with DEG and GO groupings
head(zosteraDEGroupings) #Confirm import
```

```{r}
zosteraDEGroupings <- data.frame("GeneID" = zosteraDEGroupings$seq,
                                 "name" = zosteraDEGroupings$name) #Retain GeneID and name of GO grouping. Rename seq to GeneID to match previous naming conventions
head(zosteraDEGroupings) #Confirm changes
```

```{r}
zosteraDEG.heatmap$GeneID <- rownames(zosteraDEG.heatmap) #Create GeneID column
zosteraDEGroupings.heatmap <- merge(x = zosteraDEGroupings, y = zosteraDEG.heatmap, by = "GeneID") #Merge files by gene ID
head(zosteraDEGroupings.heatmap) #Confirm merge
```

```{r}
attach(zosteraDEGroupings.heatmap) #Attach file
zosteraDEGroupings.heatmap <- zosteraDEGroupings.heatmap[order(name),] #Order by GO grouping
detach(zosteraDEGroupings.heatmap) #Detach file
zosteraDEGroupings.heatmap <- zosteraDEGroupings.heatmap[,-2] #Remove GO grouping column
zosteraDEGroupings.heatmap <- distinct(zosteraDEGroupings.heatmap) #Remove duplicate rows
rownames(zosteraDEGroupings.heatmap) <- zosteraDEGroupings.heatmap$GeneID #Set Gene ID as row names
zosteraDEGroupings.heatmap <- zosteraDEGroupings.heatmap[,-1] #Remove GeneID column
head(zosteraDEGroupings.heatmap) #Confirm changes
```

```{r}
pheatmap(zosteraDEGroupings.heatmap, cluster_row = FALSE, cluster_cols = FALSE, clustering_distance_rows = "euclidean", clustering_method = "average", show_rownames = FALSE, show_colnames = TRUE, treeheight_col = 80, legend = TRUE, color = heatmapGreen, fontsize_col = 20, fontsize_row = 12) #Create hatmap. Don't cluster rows, or columns. Use euclidean distances and average method for clustering. Do not show rownames but show column names. Adjust tree height, add a legend, and specify colors and font sizes. This looks like garbage.
```

# *L. zosterae* heatmaps

## Create color scheme

```{r}
heatmapBrown <- RColorBrewer::brewer.pal(9, "YlOrBr") #Save yellow-orange-brown color palette from RColorBrewer
```

## All genes

*Verdict: Still not exciting.*

```{r}
nZostera.GE<-read.delim("EXP.CON.nZostera.logCPM.txt", header=TRUE) #Import logCPM file
head(nZostera.GE) #Confirm import
```

```{r}
rownames(nZostera.GE) <- nZostera.GE$GeneID #Set gene ID as row names
nZostera.GE.heatmap <- nZostera.GE[,-1] #Remove redundant column
head(nZostera.GE.heatmap) #Confirm changes
```

```{r}
pheatmap(nZostera.GE.heatmap, cluster_row = TRUE, cluster_cols = FALSE, clustering_distance_rows = "euclidean", clustering_method = "average", show_rownames = FALSE, show_colnames = TRUE, treeheight_col = 80, legend = TRUE, color = heatmapBrown, fontsize_col = 20, fontsize_row = 12) #Create heatmap. Cluster rows, but not columns, Use euclidean distances and average method for clustering. Do not show rownames but show column names. Adjust tree height, add a legend, and specify colors and font sizes.
```

## Differentially expressed genes

*Verdict: Still the better one.*

```{r}
nonZosteraDEG <- read.delim("DE.EXP.CON.FDR.nZ.Annot.M.txt") #Import modified file for DEG
head(nonZosteraDEG) #Confirm import
```

I will not look at genes downregulated in the exposed treatments, so those rows should be removed.

```{r}
nonZosteraDEG <- subset(x = nonZosteraDEG, subset = nonZosteraDEG$logFC > 0) #Retain upregulated genes
nonZosteraDEG.heatmap <- merge(x = nZostera.GE, y = nonZosteraDEG, by = "GeneID") #Merge files by GeneID
head(nonZosteraDEG.heatmap) #Confirm merge
```

```{r}
rownames(nonZosteraDEG.heatmap) <- nonZosteraDEG.heatmap$GeneID #Set gene IDs as rownames
nonZosteraDEG.heatmap <- nonZosteraDEG.heatmap[,c(2:16)] #Only keep columns with logCPM for each sample
head(nonZosteraDEG.heatmap) #Confirm changes
```

```{r}
#pdf("2019-07-15-nonZostera-Differentially-Expressed-Genes-Heatmap.pdf", width = 11, height = 8.5) #Uncomment and run if you want to save the plot
pheatmap(nonZosteraDEG.heatmap, cluster_row = TRUE, cluster_cols = FALSE, clustering_distance_rows = "euclidean", clustering_method = "average", show_rownames = FALSE, show_colnames = TRUE, treeheight_col = 80, legend = TRUE, color = heatmapBrown, fontsize_col = 20, fontsize_row = 12) #Create heatmap. Cluster rows, but not columns, Use euclidean distances and average method for clustering. Do not show rownames but show column names. Adjust tree height, add a legend, and specify colors and font sizes.
#dev.off() #Uncomment and run after running plotting code above
```

