---
title: "Gene Expression Heatmaps"
author: "Yaamini Venkataraman"
date: "7/15/2019"
output: github_document
---

In this R Markdown document, I will create heatmaps for *Z. marina* and *L. zosterae* gene expression.

# Set up R Markdown document

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Install packages

```{r}
#install.packages("pheatmap") #Uncomment if you need to install this package
require(pheatmap)
```

```{r}
#install.packages("dplyr") #Uncomment if you need to install this package
require(dplyr)
```

# Obtain session information

```{r}
sessionInfo()
```

# *L. zosterae* heatmaps

## Create color scheme

```{r}
heatmapBrown <- RColorBrewer::brewer.pal(9, "YlOrBr") #Save yellow-orange-brown color palette from RColorBrewer
```

# Grace's stuff: 
## Subset genes of interest from large DE gene list

*Verdict: Still not exciting.*

```{r}
nZostera.GE<-read.delim("EXP.CON.nZostera.logCPM.txt", header=TRUE) #Import logCPM file
head(nZostera.GE) #Confirm import
```

```{r}
rownames(nZostera.GE) <- nZostera.GE$GeneID #Set gene ID as row names
nZostera.GE.heatmap <- nZostera.GE[,-1] #Remove redundant column
head(nZostera.GE.heatmap) #Confirm changes
```

Subset the genes from the non Zostera genes that are of interest
```{r}
require(dplyr)
include_list <- c("TRINITY_DN173970_c0_g1", "TRINITY_DN274805_c0_g2", "TRINITY_DN295224_c1_g1","TRINITY_DN296759_c4_g1")
nZostera.GE.heatmap.sub <- nZostera.GE.heatmap[include_list, ]
nZostera.GE.heatmap.sub
```
``` {r}

```

```{r}
#pdf("../../../../../Desktop/EIMD-2019/EWD-project/2019-07-16-nonZostera-DE-genes-of-interest.pdf", width = 11, height = 8.5)
pheatmap(nZostera.GE.heatmap.sub, cluster_row = TRUE, cluster_cols = FALSE, clustering_distance_rows = "euclidean", clustering_method = "average", show_rownames = FALSE, show_colnames = TRUE, treeheight_col = 80, legend = TRUE, color = heatmapBrown, fontsize_col = 20, fontsize_row = 12)
#dev.off()
#Create heatmap. Cluster rows, but not columns, Use euclidean distances and average method for clustering. Do not show rownames but show column names. Adjust tree height, add a legend, and specify colors and font sizes.
```

# Back to Yaamini's stuff
## Differentially expressed genes

*Verdict: Still the better one.*

```{r}
nonZosteraDEG <- read.delim("DE.EXP.CON.FDR.nZ.Annot.M.txt") #Import modified file for DEG
head(nonZosteraDEG) #Confirm import
```

I will not look at genes downregulated in the exposed treatments, so those rows should be removed.

```{r}
nonZosteraDEG <- subset(x = nonZosteraDEG, subset = nonZosteraDEG$logFC > 0) #Retain upregulated genes
nonZosteraDEG.heatmap <- merge(x = nZostera.GE, y = nonZosteraDEG, by = "GeneID") #Merge files by GeneID
head(nonZosteraDEG.heatmap) #Confirm merge
```

```{r}
rownames(nonZosteraDEG.heatmap) <- nonZosteraDEG.heatmap$GeneID #Set gene IDs as rownames
nonZosteraDEG.heatmap <- nonZosteraDEG.heatmap[,c(2:16)] #Only keep columns with logCPM for each sample
head(nonZosteraDEG.heatmap) #Confirm changes
```

### Without row numbers

```{r}
#pdf("2019-07-15-nonZostera-Differentially-Expressed-Genes-Heatmap.pdf", width = 11, height = 8.5) #Uncomment and run if you want to save the plot
pheatmap(nonZosteraDEG.heatmap, cluster_row = TRUE, cluster_cols = FALSE, clustering_distance_rows = "euclidean", clustering_method = "average", show_rownames = FALSE, show_colnames = TRUE, treeheight_col = 80, legend = TRUE, color = heatmapBrown, fontsize_col = 20, fontsize_row = 12) #Create heatmap. Cluster rows, but not columns, Use euclidean distances and average method for clustering. Do not show rownames but show column names. Adjust tree height, add a legend, and specify colors and font sizes.
#dev.off() #Uncomment and run after running plotting code above
```

### With row numbers

```{r}
nonZosteraDEG.heatmap$rownumber <- seq(from = 1, to = length(nonZosteraDEG.heatmap$S_10B), by = 1) #Add a new column numbering rows
nZgeneRowNumbers <- data.frame("GeneID" = rownames(nonZosteraDEG.heatmap),
                               "Number" = nonZosteraDEG.heatmap$rownumber) #Create new dataframe with gene IDs and row numbers
write.csv(nZgeneRowNumbers, "2019-07-15-nonZostera-Differentially-Expressed-Genes-RowNumbers.csv") #Save file
```

```{r}
rownames(nonZosteraDEG.heatmap) <- nonZosteraDEG.heatmap$rownumber #Use numbers as row names
nonZosteraDEG.heatmap <- nonZosteraDEG.heatmap[,-16] #Remove rownumber column
head(nonZosteraDEG.heatmap) #Confirm changes
```


```{r}
#pdf("2019-07-15-nonZostera-Differentially-Expressed-Genes-Rownames-Heatmap.pdf", width = 11, height = 8.5) #Uncomment and run if you want to save the plot
pheatmap(nonZosteraDEG.heatmap, cluster_row = TRUE, cluster_cols = FALSE, clustering_distance_rows = "euclidean", clustering_method = "average", show_rownames = TRUE, show_colnames = TRUE, treeheight_col = 80, legend = TRUE, color = heatmapBrown, fontsize_col = 20, fontsize_row = 5) #Create heatmap. Cluster rows, but not columns, Use euclidean distances and average method for clustering. Show row and column names. Adjust tree height, add a legend, and specify colors and font sizes.
#dev.off() #Uncomment and run after running plotting code above
```